src/Raven.Server/Integrations/PostgreSQL/PgSession.cs

```diff
@@ -0,0 +1,232 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Server.Documents;
+using Raven.Server.Integrations.PostgreSQL.Exceptions;
+using Raven.Server.Integrations.PostgreSQL.Messages;
+using Sparrow.Logging;
+
+namespace Raven.Server.Integrations.PostgreSQL
+{
+    public class PgSession
+    {
+        private static readonly Logger Logger = LoggingSource.Instance.GetLogger<PgSession>("Postgres Server");
+
+        private readonly TcpClient _client;
+        private readonly X509Certificate2 _serverCertificate;
+        private readonly int _identifier;
+        private readonly int _processId;
+        private readonly DatabasesLandlord _databasesLandlord;
+        private readonly CancellationToken _token;
+        private Dictionary<string, string> _clientOptions;
+
+        public PgSession(
+            TcpClient client,
+            X509Certificate2 serverCertificate,
+            int identifier,
+            int processId,
+            DatabasesLandlord databasesLandlord,
+            CancellationToken token)
+        {
+            _client = client;
+            _serverCertificate = serverCertificate;
+            _identifier = identifier;
+            _processId = processId;
+            _databasesLandlord = databasesLandlord;
+            _token = token;
+            _clientOptions = null;
+        }
+
+        private async Task<Stream> HandleInitialMessage(Stream stream, MessageBuilder messageBuilder)
+        {
+            var reader = PipeReader.Create(stream);
+            var writer = PipeWriter.Create(stream);
+
+            var streamToUse = stream;
+
+            var messageReader = new MessageReader();
+
+            var initialMessage = await messageReader.ReadInitialMessage(reader, _token);
+
+            if (initialMessage is SSLRequest)
+            {
+                if (_serverCertificate == null)
+                {
+                    await writer.WriteAsync(messageBuilder.SSLResponse(false), _token);
```

Comment:
What is the behavior in this case? We just reject the SSL connection?

Reply:
We say that we don't accept SSL so according to the docs - https://www.postgresql.org/docs/13/protocol-flow.html#id-1.10.5.7.11:

- "The frontend might close the connection at this point if it is dissatisfied with the response." 
- "To continue after N, send the usual StartupMessage and proceed without encryption. "